File: univariate-review/code/soy-salt-content.R
salt <- c(460, 520, 580, 700, 760, 770, 890, 910, 920, 940, 960, 1060, 1100)
mean(salt)                                  # 813.077
median(salt)                                # 890
sd(salt)                                    # 202.0885
mad(salt)                                   # 192.738
1.4826*median(abs(salt - median(salt)))     # 192.738
IQR(salt)                                   # 240 = 940 - 700 (see below)
summary(salt) 
#   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 460.0   700.0   890.0   813.1   940.0  1100.0

bitmap('../images/soy-salt-content.png', type="png256", width=7, height=7, res=250, pointsize=14) 
boxplot(salt, ylab="Salt content (mg / 15mL serving)")
dev.off()

# Testing robustness.  Create an artificial outlier
# ------------------
salt[2] = 91941
mean(salt)                                  # 7845.462   <---- not robust at all
median(salt)                                # 910        <---- robust
sd(salt)                                    # 25268.22   <---- way off!
mad(salt)                                   # 222.39     <---- robust
1.4826*median(abs(salt - median(salt)))     # 222.39     <---- manual calculation of MAD
IQR(salt)                                   # 200        <---- robust

File: univariate-review/gists/soy-salt-content.py
# Salt content, mg per 15 mL serving, for 13 soy sauces.
import numpy as np
import pandas as pd
from scipy.stats import median_abs_deviation

pd.options.plotting.backend = "plotly"

salt = pd.Series([460, 520, 580, 700, 760, 770, 890, 910,
                  920, 940, 960, 1060, 1100])

print(f"mean   = {salt.mean():.1f}")          # 813.1
print(f"median = {salt.median():.0f}")        # 890
print(f"sd     = {salt.std():.1f}")           # 202.1
# scale="normal" applies the 1.4826 factor, so the MAD is comparable
# with the standard deviation for normally distributed data.
print(f"MAD    = {median_abs_deviation(salt, scale='normal'):.1f}")   # 192.7
q1, q3 = np.percentile(salt, [25, 75])
print(f"IQR    = {q3 - q1:.0f}")              # 240 = 940 - 700

fig = salt.plot.box()
fig.update_layout(yaxis_title_text="Salt content (mg / 15mL serving)",
                  showlegend=False)
fig.show()

# Testing robustness: replace one value with an artificial outlier.
salt[1] = 91941
print(f"mean   = {salt.mean():.0f}")          # 7845, not robust at all
print(f"median = {salt.median():.0f}")        # 910, robust
print(f"sd     = {salt.std():.0f}")           # 25268, way off
print(f"MAD    = {median_abs_deviation(salt, scale='normal'):.1f}")   # 222.4
